test/test_script/test_external_m3u_playlist.cc

```diff
@@ -0,0 +1,176 @@
+#include <cds_objects.h>
+#include <duktape.h>
+#include <string_converter.h>
+#include <cds_objects.h>
+#include <memory>
+
+#include "mock/common_script_mock.h"
+#include "mock/duk_helper.h"
+#include "mock/script_test_fixture.h"
+
+using namespace ::testing;
+using namespace std;
+
+class ExternalUrlM3UPlaylistTest : public ScriptTestFixture {
+ public:
+  ExternalUrlM3UPlaylistTest() {
+    commonScriptMock.reset(new ::testing::NiceMock<CommonScriptMock>());
+    scriptName = "playlists.js";
+  };
+
+  virtual ~ExternalUrlM3UPlaylistTest() {
+    commonScriptMock.reset();
+  };
+  // As Duktape requires static methods, so must the mock expectations be
+  static unique_ptr<CommonScriptMock> commonScriptMock;
+
+  // Used to iterate through `readln` content
+  static int readLineCnt;
+};
+
+unique_ptr<CommonScriptMock> ExternalUrlM3UPlaylistTest::commonScriptMock;
+int ExternalUrlM3UPlaylistTest::readLineCnt = 0;
+
+static duk_ret_t getPlaylistType(duk_context *ctx) {
+  string playlistMimeType = ScriptTestFixture::getPlaylistType(ctx);
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->getPlaylistType(playlistMimeType);
+}
+
+static duk_ret_t print(duk_context *ctx) {
+  string msg = ScriptTestFixture::print(ctx);
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->print(msg);
+}
+
+static duk_ret_t createContainerChain(duk_context *ctx) {
+  vector<string> array = ScriptTestFixture::createContainerChain(ctx);
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->createContainerChain(array);
+}
+
+static duk_ret_t getLastPath(duk_context* ctx) {
+  string inputPath = ScriptTestFixture::getLastPath(ctx);
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->getLastPath(inputPath);
+}
+
+// Proxy the Duktape script with `readln`
+// Mimics reading the playlist file line by line
+// Uses the `CommonScriptMock` to track expectations
+static duk_ret_t readln(duk_context *ctx) {
+  vector<string> lines = {
+    "#EXTM3U",
+    "#EXTINF:-1,(#1 - 177/750) : Ibiza Global Radio :",
+    "http://46.105.171.217:8024",
+    "-EOF-"  // used to stop processing
+  };
+
+  string line = lines.at(ExternalUrlM3UPlaylistTest::readLineCnt);
+
+  duk_push_string(ctx, line.c_str());
+  ExternalUrlM3UPlaylistTest::readLineCnt++;
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->readln(line);
+}
+
+static duk_ret_t addCdsObject(duk_context *ctx) {
+  vector<string> keys = {"mimetype", "objectType", "location",
+                         "title", "protocol", "upnpclass",
+                         "description", "playlistOrder" };
+  addCdsObjectParams params = ScriptTestFixture::addCdsObject(ctx, keys);
+  return ExternalUrlM3UPlaylistTest::commonScriptMock->addCdsObject(params.objectValues, params.containerChain, params.objectType);
+}
+
+// Mock the Duktape C methods
+// that are called from the playlists.js script
+// * These are static methods, which makes mocking difficult.
+static  duk_function_list_entry js_global_functions[] =
+{
```

Comment:
Each test overrides the global JS function with its own set..to verify behavior.  Ultimately ALL script behavior ends up funnelling into c++ function in the real implementation.  This way we can capture the data at the c++ funnel point and verify it.  This confirms the script did whatever it wanted, but still produced the expected result.